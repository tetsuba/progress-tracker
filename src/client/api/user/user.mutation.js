import gql from 'graphql-tag'

export const VERIFY_EMAIL_MUTATION = gql`
  mutation($input: UserEmailInput!) {
    verifyEmail(input: $input) {
      confirmation
    }
  }
`

export const SEND_PASSWORD_RESET_CONFIRMATION_MUTATION = gql`
  mutation($input: UserEmailInput!) {
    sendPasswordResetConfirmation(input: $input) {
      confirmation
    }
  }
`

export const LOGIN_MUTATION = gql`
  mutation($input: LoginUserInput!) {
    userLogin(input: $input) {
      id
      firstName
      lastName
      token
      email
    }
  }
`

export const REGISTER_USER_MUTATION = gql`
  mutation NewUser($input: NewUserInput!) {
    newUser(input: $input) {
      id
    }
  }
`

export const REST_PASSWORD_MUTATION = gql`
  mutation ResetPassword($input: ResetPasswordInput!) {
    resetPassword(input: $input) {
      confirmation
    }
  }
`
