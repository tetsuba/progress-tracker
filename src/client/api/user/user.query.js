import gql from 'graphql-tag'

export const GET_USER_DETAILS_QUERY = gql`
  {
    getUserDetails {
      firstName
      lastName
      email
    }
  }
`

export const IS_USER_SESSION_EXPIRED = gql`
  {
    isUserSessionExpired {
      success
    }
  }
`

export const VALIDATE_USER_EMAIL_QUERY = gql`
  query ValidateUserEmail($token: String!) {
    validateUserEmail(token: $token) {
      success
    }
  }
`
