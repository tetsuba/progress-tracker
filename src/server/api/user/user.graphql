
type UserType {
    id: String
    firstName: String
    lastName: String
    email: String
    password: String
    country: String
}

type UserLoginType {
    id: String
    firstName: String
    lastName: String
    token: String
    email: String
}

type UserVerifyType {
    confirmation: String
}

input NewUserInput {
    firstName: String
    lastName: String
    email: String
    password: String
    country: String
}

input LoginUserInput {
    email: String
    password: String
}

input UserEmailInput {
    email: String
}

input UserDataInput {
    id: String
    firstName: String
    lastName: String
}

input ResetPasswordInput {
    id: String
    password: String
}

extend type Query {
    getUserData: UserType!
    isUserSessionExpired(token: String): UserType!
}

type Mutation {
    newUser(input: NewUserInput!): UserType!
    userLogin(input: LoginUserInput!): UserLoginType!
    updateUserData(input: UserDataInput!): UserLoginType!
    verifyEmail(input: UserEmailInput!): UserVerifyType!
    sendPasswordResetConfirmation(input: UserEmailInput!): UserVerifyType!
    resetPassword(input: ResetPasswordInput!): UserVerifyType!
}
