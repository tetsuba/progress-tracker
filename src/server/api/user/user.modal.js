const mongoose = require('mongoose')

const userSchema = new mongoose.Schema(
  {
    firstName: {
      type: String,
      required: true,
    },
    lastName: {
      type: String,
      required: true,
    },
    email: {
      type: String,
      required: true,
    },
    password: {
      type: String,
      required: true,
    },
    country: {
      type: String,
      required: false,
    },
    isVerified: {
      type: Boolean,
      default: false,
    },
  },
  { _id: true }
)

module.exports = userSchema
