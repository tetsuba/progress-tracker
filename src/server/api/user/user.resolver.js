const { validateEmail } = require('../../utils/common')

const {
  createNewUser,
  findUser,
  updateUser,
  findEmail,
  resetPassword,
} = require('./user.CRUD')
const { getEmailMailOptions } = require('../../utils/email/sendEmail.js')
const { errorName } = require('../../errorHandling')
const { createVerificationToken } = require('../../utils/token')
const { passwordsDoNotMatch } = require('../utils')

module.exports = {
  Query: {
    isUserSessionExpired: async (_, args, context) => {
      try {
        return await findUser({
          email: context.user.email,
          password: context.user.password,
        })
      } catch (e) {
        return e
      }
    },
    getUserData: async (_, args, context) => {
      console.log('@@@@@@@@@@@@ getUserData: ', context)
      try {
        return await findUser({
          email: context.user.email,
          password: context.user.password,
        })
      } catch (e) {
        return e
      }
    },
  },

  Mutation: {
    newUser: async (_, args, context, info) => {
      const name = 'confirmEmail'

      if (validateEmail(args.input.email)) {
        throw new Error(errorName.NOT_VALID_EMAIL)
      }

      try {
        const mailOptions = getEmailMailOptions(name)
        return await createNewUser(args.input, mailOptions)
      } catch (err) {
        return err
      }
    },

    userLogin: async (_, args, context, info) => {
      const { User, Token } = context.models
      const {
        input: { email, password },
      } = args
      const user = await User.findOne({ email })

      /* Error Handling:
       * - User is not found
       * - Passwords do not match
       *  */
      if (!user || passwordsDoNotMatch(password, user.password)) {
        throw new Error(errorName.INCORRECT_USER_DETAILS)
      }

      /* Error Handling:
       * - User email has not been verified
       *  */
      if (!user.isVerified) {
        throw new Error(errorName.EMAIL_NOT_VERIFIED)
      }

      /* Create a session token (12hrs):
       *  */
      const verificationToken = createVerificationToken(user)
      const { token } = await Token.create({
        _userId: user._id,
        token: verificationToken,
      })

      return { token }
    },

    updateUserData: async (_, args, context, info) => {
      try {
        return await updateUser(args.input)
      } catch (err) {
        return err
      }
    },

    verifyEmail: async (_, args, context) => {
      const name = 'confirmEmail'
      try {
        const mailOptions = getEmailMailOptions(name)
        return await findEmail({ email: args.input.email }, mailOptions)
      } catch (err) {
        if (err.name === 'TypeError') {
          console.log(
            `Property "name": "${name}" is incorrect and can not get mail options`
          )
          return new Error(errorName.INTERNAL_SERVER_ERROR)
        }
        return err
      }
    },

    sendPasswordResetConfirmation: async (_, args, context) => {
      const name = 'resetPassword'
      try {
        const mailOptions = getEmailMailOptions(name)
        return await findEmail({ email: args.input.email }, mailOptions)
      } catch (err) {
        if (err.name === 'TypeError') {
          console.log(
            `Property "name": "${name}" is incorrect and can not get mail options`
          )
          return new Error(errorName.INTERNAL_SERVER_ERROR)
        }
        return err
      }
    },

    resetPassword: async (_, args, context) => {
      try {
        return await resetPassword(args.input.token, args.input.password)
      } catch (err) {
        return err
      }
    },
  },
}

// NOTES
// fieldName: (parent, args, context, info) => data;

// parent:  An object that contains the result returned from the resolver on the parent type
// args:    An object that contains the arguments passed to the field
// context: An object shared by all resolvers in a GraphQL operation. We use the context to contain per-request state such as authentication information and access our data sources.
// info:    Information about the execution state of the operation which should only be used in advanced cases
