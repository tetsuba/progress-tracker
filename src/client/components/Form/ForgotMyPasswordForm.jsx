import React from 'react';
import { useMutation } from '@apollo/react-hooks';
import { Button, Form, Row, Col } from 'react-bootstrap';

// MUTATIONS
import { SEND_PASSWORD_RESET_CONFIRMATION_MUTATION } from '../../api/user/user.mutation';

// ICONS
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {faCaretLeft} from "@fortawesome/free-solid-svg-icons";

// COMPONENT
import TextLink from "../TextLink/TextLink";
import {useError, useInputChange, useSuccess} from "../../hooks/hooks";

const ForgotMyPasswordForm = ({ defaultEmail = '', resetPassword}) => {
    const [ sendPasswordResetConfirmation ] = useMutation(SEND_PASSWORD_RESET_CONFIRMATION_MUTATION);
    const [inputs, setInputs] = useInputChange({ email: defaultEmail });
    const [error, setError] = useError({ message: '' });
    const [success, setSuccess] = useSuccess('sendPasswordResetConfirmation');
    const options = { variables: { input: inputs } };

    return success
        ? (
            <Row>
                <h3>Please check your email.</h3>
            </Row>
        )
        :(
            <Col>
                <Row>
                    <h3>Forgot my password</h3>
                </Row>
                <Row>
                    <p>Enter your email address and we will send you a link to reset your password.</p>
                </Row>
                <Row>
                    <Form
                        id="ForgotMyPasswordForm"
                        className="w-100"
                        onSubmit={(e) => {
                            e.preventDefault();
                            sendPasswordResetConfirmation(options)
                                .then(setSuccess)
                                .catch(setError)
                        }}
                    >
                        <Form.Group controlId="forgotMyPasswordEmail">
                            <Form.Control
                                required
                                type="email"
                                placeholder="Enter email"
                                name="email"
                                onChange={setInputs}
                                value={inputs.email}
                                isInvalid={!!error.message}
                            />
                            <Form.Control.Feedback type="invalid">
                                {error.message}
                            </Form.Control.Feedback>
                            <Form.Text className="text-muted">
                                We'll never share your email with anyone else.
                            </Form.Text>
                        </Form.Group>
                        <Button id="ForgotMyPasswordSubmit" variant="primary" type="submit">
                            Reset Password
                        </Button>
                    </Form>
                </Row>
                { resetPassword && (
                    <Row className="mt-3">
                        <TextLink eventHandler={resetPassword}>
                            <FontAwesomeIcon icon={faCaretLeft} /> back
                        </TextLink>
                    </Row>
                )}

            </Col>
        )
};

export default ForgotMyPasswordForm;
