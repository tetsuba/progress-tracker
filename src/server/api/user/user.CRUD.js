const { getUserId, deleteToken } = require('../token/token.CRUD');
const jwt = require('jsonwebtoken');

// This order can not be changed
const mongoose = require('mongoose');
require('./user.modal');
const User = mongoose.model('user');
// -----------------------------

const { createToken } = require('../token/token.CRUD');
const { sendMail } = require('../../utils/email/sendEmail');
const { errorName } = require('../../errorHandling');
const { createAuthToken } = require('../../utils/token');

async function findEmail(data, mailOptions) {
    try {
        const user = await User.findOne({ email: data.email });

        // ERROR - User not found
        if (!user) {
            throw new Error(errorName.EMAIL_DOES_NOT_EXIST)
        }

        // ERROR - User is verified
        if (user.isVerified && mailOptions.subject.includes('confirm')) {
            throw new Error(errorName.EMAIL_VERIFIED);
        }

        // Do not send an email when running functional tests
        if (!process.env.TEST) {
            const obj = await createToken(user);
            await sendMail(obj.token, mailOptions);
        }

        return {
            confirmation: 'Please check your email and confirm.'
        }

    } catch(error) {
        return error
    }

}

async function findUser(data) {
    try {
        const user = await User.findOne({ email: data.email });

        // ERROR - User is not found or password does not match
        if (!user || user.password !== data.password) {
            throw new Error(errorName.INCORRECT_USER_DETAILS)
        }

        // ERROR - Email address not verified
        if (!user.isVerified) {
            throw new Error(errorName.EMAIL_NOT_VERIFIED)
        }

        const token = createAuthToken(user);

        const payload = jwt.verify(
            token,
            process.env.REACT_APP_JWT_AUTH_SECRET
    )

        return {
            firstName: user.firstName,
            lastName: user.lastName,
            email: user.email,
            id: user._id,
            token
        }
    } catch(error) {
        return error
    }
}

// RegisterUser
async function createNewUser(data) {
    try {
        const user = await User.findOne({ email: data.email});

        // ERROR - User is not found or password does not match
        if (user) {
            throw new Error(errorName.EMAIL_ALREADY_EXIST)
        }

        const newUser = await User.create(data);
        const obj = await createToken(newUser);
        sendMail(obj.token);
        return { success: 'success' };
    } catch(err) {
        return err
    }
}

async function updateUser(data) {
    console.log('updateUser:', data);
    try {
        const user = await User.findById(data.id);
        console.log('User:', user);
        user.firstName = data.firstName;
        user.lastName = data.lastName;
        return await user.save()
    } catch (e) {
        console.log('findUser error', e);
        return e;
    }
}

async function userVerified(id) {
    try {
        const user = await User.findById(id);
        user.isVerified = true;
        return await user.save()
    } catch (e) {
        console.log('userVerified error', e);
        return e;
    }
}

async function resetPassword(token, password) {
    try {
        const id = await getUserId(token);
        const user = await User.findById(id);
        user.password = password;
        await user.save();

        // Don not delete token when running functional tests
        if (!process.env.TEST) {
            await deleteToken(token);
        }

        return {
            confirmation: 'New password is saved'
        }
    } catch (e) {
        console.log('resetPassword error', e);
        return e;
    }
}

module.exports = {
    findUser,
    createNewUser,
    updateUser,
    userVerified,
    findEmail,
    resetPassword,
};
