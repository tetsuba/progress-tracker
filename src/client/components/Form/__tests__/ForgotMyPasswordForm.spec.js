import { GraphQLError } from "graphql";
import { act } from 'react-dom/test-utils';
import { graphRenderer, delay } from "../../../../test/testHelper";

// COMPONENTS
import ForgotMyPasswordForm from "../ForgotMyPasswordForm";

// MUTATIONS
import { SEND_PASSWORD_RESET_CONFIRMATION_MUTATION}  from "../../../api/user/user.mutation";

describe('<ForgotMyPasswordForm>', () => {

    const request = {
        query: SEND_PASSWORD_RESET_CONFIRMATION_MUTATION,
        variables: { input: { email: 'test@test.com'} },
    };

    const id = '#ForgotMyPasswordForm';
    const props = {
        defaultEmail: 'test@test.com',
        resetPassword: jest.fn(),
    };

    describe('@Render', () => {
        const result = {
            data: { sendPasswordResetConfirmation: { confirmation: 'Confirm'} }
        };

        it('should render email password reset form', () => {
            const mocks = [{ request, result }];
            const wrapper = graphRenderer(ForgotMyPasswordForm, mocks, props);
            expect(wrapper.find(ForgotMyPasswordForm)).toMatchSnapshot();
        });

        it('should render a success message', async() => {
            const mocks = [{ request, result }];
            const wrapper = graphRenderer(ForgotMyPasswordForm, mocks, props);

            act(() => {
                wrapper.find(id).get(0).props.onSubmit({
                    preventDefault: jest.fn(),
                });
            });

            await delay();
            wrapper.update();
            expect(wrapper.find(ForgotMyPasswordForm)).toMatchSnapshot();
        });
    });

    describe('@Error', () => {
        const errorMessage = 'error occurred';
        const errors = [new GraphQLError(errorMessage)];

        it('should render an error message if form returns an error', async() => {
            const mocks = [{ request, result: {errors}}];
            const wrapper = graphRenderer(ForgotMyPasswordForm, mocks, props);

            act(() => {
                wrapper.find(id).get(0).props.onSubmit({
                    preventDefault: jest.fn(),
                });
            });

            await delay();
            wrapper.update();
            wrapper.find({type: 'invalid'}).forEach((message) => {
                expect(message.prop('children')).toEqual(errorMessage)
            })
        })
    })
});