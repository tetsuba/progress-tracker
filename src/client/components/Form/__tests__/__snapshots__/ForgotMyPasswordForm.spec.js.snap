// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ForgotMyPasswordForm> @Render should render 1`] = `
<Box
  max={500}
>
  <Row
    noGutters={false}
  >
    <Col>
      <Form
        className="w-100"
        id="ForgotMyPasswordForm"
        inline={false}
        onSubmit={[Function]}
      >
        <FormGroup
          controlId="ForgotMyPasswordEmail"
        >
          <FormControl
            isInvalid={false}
            name="email"
            onChange={[Function]}
            placeholder="@"
            required={true}
            type="email"
            value=""
          />
          <Feedback
            type="invalid"
          />
          <FormText
            className="text-muted"
          >
            We'll never share your email with anyone else.
          </FormText>
        </FormGroup>
        <Button
          active={false}
          className="float-right"
          disabled={false}
          id="ForgotMyPasswordSubmit"
          type="submit"
          variant="primary"
        >
          Reset Password
        </Button>
      </Form>
    </Col>
  </Row>
</Box>
`;

exports[`<ForgotMyPasswordForm> @Render should render back button 1`] = `
<Box
  max={500}
>
  <Row
    noGutters={false}
  >
    <Col>
      <Form
        className="w-100"
        id="ForgotMyPasswordForm"
        inline={false}
        onSubmit={[Function]}
      >
        <FormGroup
          controlId="ForgotMyPasswordEmail"
        >
          <FormControl
            isInvalid={false}
            name="email"
            onChange={[Function]}
            placeholder="@"
            required={true}
            type="email"
            value=""
          />
          <Feedback
            type="invalid"
          />
          <FormText
            className="text-muted"
          >
            We'll never share your email with anyone else.
          </FormText>
        </FormGroup>
        <Button
          active={false}
          className="float-right"
          disabled={false}
          id="ForgotMyPasswordSubmit"
          type="submit"
          variant="primary"
        >
          Reset Password
        </Button>
      </Form>
      <TextLink
        eventHandler={[MockFunction]}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                192,
                512,
                Array [],
                "f0d9",
                "M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z",
              ],
              "iconName": "caret-left",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
         back
      </TextLink>
    </Col>
  </Row>
</Box>
`;
